### PR TITLE
Add Legendre symbol.

### PR DESCRIPTION
**Pull request by mmaker**

### CODE REVIEW

**Comment by mmaker** [comment]
oops, there's a bunch of whitespace fixes that involuntarily ended up in the patch. 
(I'm not sure what's your policy on them, is it still okay?)

**Comment by ebfull** [comment]
> (I'm not sure what's your policy on them, is it still okay?)

Definitely okay!

File: src/bls12_381/fr.rs [hunk 1]
```diff
@@ -551,49 +552,53 @@ impl Fr {
 }
 
 impl SqrtField for Fr {
+
+    fn legendre(&self) -> ::LegendreSymbol {
+        let s = self.pow([0x7fffffff80000000, 0xa9ded2017fff2dff, 0x199cec0404d0ec02, 0x39f6d3a994cebea4]);
```
Comment (ebfull) [inline src/bls12_381/fr.rs]: Label that this is (r-1)/2.

**Comment by ebfull** [comment]
One slight improvement: in the `random_sqrt_tests` tests in `src/tests/field.rs`, we could randomly generate quadratic residues and ensure the Legendre symbol is correct.

**Comment by ebfull** [comment]
Other than that, I'm +1 on this patch. Thank you! 🎉

File: src/bls12_381/fr.rs (hunk 1 above)
Comment (mmaker) [inline src/bls12_381/fr.rs]: done!

**Comment by mmaker** [comment]
The two changes should be in the attached commit.

**Comment by ebfull** [comment]
Excellent! Thank you!

@bmerge r+

**Comment by bmerge** [comment]
:pushpin: Commit 874fc45 has been approved by `ebfull`

**Comment by bmerge** [comment]
:hourglass: Testing commit 874fc45a307430d1a8da106251ea63fed685a6d2 with merge 96410bd9d7e785f001184b63711c6a89028b4312...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-auto](http://ci.bellman.rs:8010/builders/pairing-linux64-auto/builds/23)

**Comment by mmaker** [comment]
herp derp, I thought "sqrt" would have matched that test as well.

In that branch of code we don't know which one between b or negb, plus (for instance) I incorrectly assumed "r" is a safe prime. I quickly amended the commit removing the invalid check.

**Comment by bmerge** [comment]
@mmaker: :key: Insufficient privileges: Not in reviewers

**Comment by mmaker** [comment]
héhéhéhé

**Comment by bmerge** [comment]
@mmaker: :key: Insufficient privileges: Not in reviewers

**Comment by ebfull** [comment]
@bmerge r+

**Comment by bmerge** [comment]
:pushpin: Commit 18f02ff has been approved by `ebfull`

**Comment by bmerge** [comment]
:hourglass: Testing commit 18f02ff33b0e1031b7e8e5187ae32822f11a6c4b with merge 65b92aa9cfc63817f338c7f56fb30fec8a7fd09f...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-auto](http://ci.bellman.rs:8010/builders/pairing-linux64-auto/builds/24)

**Comment by mmaker** [comment]
this time it is: 
```
error: this expression borrows a reference that is immediately dereferenced by the compiler
   --> src/bls12_381/fq2.rs:161:19
    |
161 |         Fq2::norm(&self).legendre()
    |                   ^^^^^
```
once again I've explicitly put self as argument rather than doing `self.method`… my apologies.
once again, I amended and force-pushed on my private repository.

**Comment by ebfull** [comment]
@bmerge r+

**Comment by bmerge** [comment]
:pushpin: Commit 9846ad2 has been approved by `ebfull`

**Comment by bmerge** [comment]
:hourglass: Testing commit 9846ad2d177863c03f524a79aff04bb2f7734ce1 with merge 57de78f4f1052fb6adbc201914158711986ed9b5...

**Comment by bmerge** [comment]
:sunny: Test successful - [pairing-linux32-auto](http://ci.bellman.rs:8010/builders/pairing-linux32-auto/builds/23), [pairing-linux64-auto](http://ci.bellman.rs:8010/builders/pairing-linux64-auto/builds/25), [pairing-windows32msvc-auto](http://ci.bellman.rs:8010/builders/pairing-windows32msvc-auto/builds/23), [pairing-windows64msvc-auto](http://ci.bellman.rs:8010/builders/pairing-windows64msvc-auto/builds/23)
Approved by: ebfull
Pushing 57de78f4f1052fb6adbc201914158711986ed9b5 to master...

**Pull request merged**